Fix duplicate extension loading issue described in #1016

guobingkun opened the pull request:
This patch aims to fix issue #1016  

Currently if we put something in ```druid.extensions.coordinates```, the local module that was previsouly loaded will be accidentally loaded again. The reason is that the return value of the second ```ServiceLoader.load(clazz, loader)``` in the original code still contains previously loaded module. Since ```Set<T> retVal``` is not doing its job by filtering out loaded modules, previously loaded module will be added again. If that module happens to create Guice bindings, duplicate binding errors will occur.

This patch adds a HashSet that keeps track the module names that have already been loaded.

I also rearranged the order in which the extensions are loaded. Extensions explicitly specified in ```druid.extensions.coordinates``` have a higher priority.

guobingkun commented:
@drcrallen

xvrl commented:
Thanks @guobingkun do you know if we can test this somehow, either unit test or integration test?
It would also be nice to update the documentation to explain precedence of extensions over classpath.

himanshug commented:
@guobingkun  LGTM 

+1 for some test, UT should be easy.

guobingkun commented:
Hi @xvrl @himanshug I added UT and updated some doc. Notice that I created a test-only getter for ```loadersMap``` so that UT will bypass ```aether.resolveArtifacts(dependencyRequest)```

xvrl commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -118,33 +119,46 @@
+    final Set<String> extensionNames = Sets.newHashSet();
 
     for (String coordinate : config.getCoordinates()) {
       log.info("Loading extension[%s] for class[%s]", coordinate, clazz.getName());
```
We should probably use `getCanonicalName` to keep things clean and prevent weird things like anonymous or local classes. That means we also should failgracefully if `getCanonicalName` returns `null`.

guobingkun replied:
Hi @xvrl , I made some changes so that it ignores extension modules that don't have a canonical class name, and gives a log warn message.

xvrl replied:
@guobingkun ok, is there a reason not to use the canonical name instead of getName()? I feel the canonical name is more user-friendly for logging. I don't think that should break anything.

guobingkun replied:
@xvrl I thought the ```clazz``` parameter passed to this function is never an anonymous or local class. Are you suggesting we should ignore the entire function if ```clazz``` doesn't have a canonical name? Do you want to add that check for the extension modules as well?

xvrl replied:
no, all I am saying is that we should replace
`log.info("Loading extension[%s] for class[%s]", coordinate, clazz.getName());`
with
`log.info("Loading extension[%s] for class[%s]", coordinate, clazz.getCanonicalName());`

xvrl replied:
nevermind @guobingkun, I think we're ok here, it's probably not necessary to use canonical name for logging everywhere if that forces us to add checks.

guobingkun replied:
Sure, do you still want the 3rd commit here?

xvrl replied:
@guobingkun yes, leave things as is

drcrallen commented on server/src/main/java/io/druid/initialization/Initialization.java:
```diff
@@ -118,33 +119,46 @@
+  /**
+   * Used for testing only
+   */
+  static Map<String, URLClassLoader> getLoadersMap()
```
Can this be `protected` then if it is only for UT?

xvrl replied:
It has to be package protected, for the test class to be able to access it, so this is fine.

drcrallen replied:
For some reason I was reading "public" when it very clearly isn't there

xvrl commented:
@guobingkun can you squash your commits, and I'll merge.

guobingkun commented:
@xvrl Done.
